### PR TITLE
Implement UnormR8G8B8A8 vertex format.

### PR DESCRIPTION
**Pull request by SenorBlanco**
This likely requires a test in InputStateTests, but I'm not sure exactly how. Suggestions?

### CODE REVIEW

File: src/tests/end2end/InputStateTests.cpp [hunk 1]
```diff
@@ -51,6 +51,8 @@ class InputStateTest : public NXTTest {
                     return component >= 2;
                 case VertexFormat::FloatR32:
                     return component >= 1;
+                case VertexFormat::UnormR8G8B8A8:
+                    return component >= 4;
```
Comment (kainino0x) [inline src/tests/end2end/InputStateTests.cpp]: nit: this case could be merged with FloatR32G32B32A32

File: src/backend/InputState.cpp [hunk 2]
```diff
@@ -42,6 +42,8 @@ namespace backend {
                 return 2;
             case nxt::VertexFormat::FloatR32:
                 return 1;
+            case nxt::VertexFormat::UnormR8G8B8A8:
```
Comment (Kangz) [inline src/backend/InputState.cpp]: Could be merged with FloatR32G32B32A32

File: next.json [hunk 3]
```diff
@@ -1209,7 +1209,8 @@
             {"value": 0, "name": "float r32 g32 b32 a32"},
             {"value": 1, "name": "float r32 g32 b32"},
             {"value": 2, "name": "float r32 g32"},
-            {"value": 3, "name": "float r32"}
+            {"value": 3, "name": "float r32"},
+            {"value": 4, "name": "unorm r8 g8 b8 a8"}
```
Comment (Kangz) [inline next.json]: Can this add the 2 and 1 component versions too?

**Comment by Kangz** [comment]
Yeah, let's skip the test for now.

**Comment by SenorBlanco** [comment]
OK, I implemented the unorm 1- and 2-component formats for completeness. Note that the Metal
headers seem to be missing MTLVertexFormatUCharNormalized, but the docs have it. Not sure what's going on there, but I used the enum value from the docs (ugh).

**Comment by SenorBlanco** [comment]
OK, updating to XCode9 fixed it. Using proper enum value now.

File: src/backend/metal/InputStateMTL.mm [hunk 4]
```diff
@@ -30,6 +30,12 @@ MTLVertexFormat VertexFormatType(nxt::VertexFormat format) {
                     return MTLVertexFormatFloat2;
                 case nxt::VertexFormat::FloatR32:
                     return MTLVertexFormatFloat;
+                case nxt::VertexFormat::UnormR8G8B8A8:
+                    return MTLVertexFormatUChar4Normalized;
+                case nxt::VertexFormat::UnormR8G8:
+                    return MTLVertexFormatUChar2Normalized;
+                case nxt::VertexFormat::UnormR8:
+                    return MTLVertexFormatUCharNormalized;
```
Comment (Kangz) [inline src/backend/metal/InputStateMTL.mm]: Looks like this is only available on more recent Apple systems, can we skip UnormR8 for now? Sorry for the churn.

File: src/backend/InputState.cpp [hunk 5]
```diff
@@ -35,30 +35,41 @@ namespace backend {
...
             case nxt::VertexFormat::FloatR32:
+            case nxt::VertexFormat::UnormR8:
                 return 1;
             default:
                 UNREACHABLE();
         }
     }
 
-    size_t VertexFormatSize(nxt::VertexFormat format) {
+    size_t VertexFormatComponentSize(nxt::VertexFormat format) {
```
Comment (Kangz) [inline src/backend/InputState.cpp]: Can this be declared in the InputState.h header?

**Comment by Kangz** [comment]
Still LGTM, the CI error is caused by UnormR8 not being available yet on the Metal version Travis has. Also can you squash all commits?

**Comment by Kangz** [comment]
Thank you!

**Pull request merged**